gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb

```diff
@@ -0,0 +1,823 @@
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "search_kaggle_movies(\"Harry Potter\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define agent\n",
+    "\n",
+    "The third component of your agent involves adding a reasoning layer, which helps your agent use the tools that you provided to help the end user achieve a higher-level goal.\n",
+    "\n",
+    "If you were to use Gemini and Function Calling on their own without a reasoning layer, you would need to handle the process of calling functions and APIs in your application code, and you would need to implement retries and additional logic to ensure that your function calling code is resilient to failures and malformed requests.\n",
+    "\n",
+    "Define the prompt template and initialize the chat session history:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define prompt template\n",
+    "prompt = {\n",
+    "    \"history\": lambda x: x[\"history\"],\n",
+    "    \"input\": lambda x: x[\"input\"],\n",
+    "    \"agent_scratchpad\": (\n",
+    "        lambda x: format_to_openai_function_messages(x[\"intermediate_steps\"])\n",
+    "    ),\n",
+    "} | prompts.ChatPromptTemplate.from_messages(\n",
+    "    [\n",
+    "        (\"placeholder\", \"{history}\"),\n",
+    "        (\"user\", \"{input}\"),\n",
+    "        (\"placeholder\", \"{agent_scratchpad}\"),\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "# Initialize session history\n",
+    "store = {}\n",
+    "\n",
+    "\n",
+    "def get_session_history(session_id: str):\n",
```

Comment:
Can you add a return type for this method?

Reply:
Done!